src/rt/ds/stackarray.h

```diff
@@ -0,0 +1,63 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include <cassert>
+#include <cstddef>
+
+/**
+ * @brief A stack allocated array.
+ *
+ * Due to portability allocates a largeish array on the stack, if this array is
+ * not big enough then dynamically allocates something of the correct size. This
+ * is done to avoid the need for a dynamic allocation in the common case.
+ */
+template<typename T>
+class StackArray
+{
+  // Size of the stack allocated array.
+  static constexpr size_t Size = 128;
+
+  // Stack allocated array.
+  T main[Size];
+
+  // Pointer to the array in use, may be main or a dynamically allocated array.
+  T* current;
+
+  // User requested size.
+  std::size_t size;
+
+public:
+  StackArray(std::size_t size) : current(main), size(size)
+  {
+    if (size > Size)
```

Comment:
You could use `std::size(main)` instead of `Size` to express a bit more precisely which size you are comparing against.

Reply:
This comment will become invalid if main is changed in type from T for initialization purposes.

Reply:
I switched to the untyped memory to fix the constructor issues.